libc/src/__support/CPP/bit.h

```diff
@@ -1,57 +1,261 @@
+/// Returns cpp::numeric_limits<T>::digits on an input of all ones.
+template <typename T>
+[[nodiscard]] LIBC_INLINE constexpr int countr_one(T value) {
+  static_assert(cpp::is_unsigned_v<T>);
+  return cpp::countr_zero<T>(~value);
+}
+
+/// Returns the number of bits needed to represent value if value is nonzero.
+/// Returns 0 otherwise.
+///
+/// Ex. bit_width(5) == 3.
+template <typename T>
+[[nodiscard]] LIBC_INLINE constexpr int bit_width(T value) {
+  static_assert(cpp::is_unsigned_v<T>);
+  return cpp::numeric_limits<T>::digits - cpp::countl_zero(value);
+}
+
+/// Returns the largest integral power of two no greater than value if value is
+/// nonzero.  Returns 0 otherwise.
+///
+/// Ex. bit_floor(5) == 4.
+template <typename T> [[nodiscard]] LIBC_INLINE constexpr T bit_floor(T value) {
+  static_assert(cpp::is_unsigned_v<T>);
+  if (!value)
+    return 0;
+  return T(1) << (cpp::bit_width(value) - 1);
+}
+
+/// Returns the smallest integral power of two no smaller than value if value is
+/// nonzero.  Returns 1 otherwise.
+///
+/// Ex. bit_ceil(5) == 8.
+///
+/// The return value is undefined if the input is larger than the largest power
+/// of two representable in T.
+template <typename T> [[nodiscard]] LIBC_INLINE constexpr T bit_ceil(T value) {
+  static_assert(cpp::is_unsigned_v<T>);
+  if (value < 2)
+    return 1;
+  return T(1) << cpp::bit_width<T>(value - 1u);
+}
+
+/// Count the number of set bits in a value.
+/// Ex. popcount(0xF000F000) = 8
+/// Returns 0 if the word is zero.
+template <typename T, typename = cpp::enable_if_t<cpp::is_unsigned_v<T>>>
+[[nodiscard]] LIBC_INLINE constexpr int popcount(T value) {
+  if constexpr (sizeof(T) == 8) {
+#if LIBC_HAS_BUILTIN(__builtin_popcountll)
+    return (int)__builtin_popcountll(value);
+#endif
+    uint64_t v = value;
+    v = v - ((v >> 1) & 0x5555555555555555ULL);
+    v = (v & 0x3333333333333333ULL) + ((v >> 2) & 0x3333333333333333ULL);
+    v = (v + (v >> 4)) & 0x0F0F0F0F0F0F0F0FULL;
+    return int((uint64_t)(v * 0x0101010101010101ULL) >> 56);
+  } else if constexpr (sizeof(T) <= 4) {
+#if LIBC_HAS_BUILTIN(__builtin_popcount)
+    return (int)__builtin_popcount(value);
+#endif
+    uint32_t v = value;
+    v = v - ((v >> 1) & 0x55555555);
+    v = (v & 0x33333333) + ((v >> 2) & 0x33333333);
+    return int(((v + (v >> 4) & 0xF0F0F0F) * 0x1010101) >> 24);
+  } else {
+    static_assert(cpp::always_false<T>);
+  }
+}
+
+// Forward-declare rotr so that rotl can use it.
+template <typename T, typename = cpp::enable_if_t<cpp::is_unsigned_v<T>>>
+[[nodiscard]] LIBC_INLINE constexpr T rotr(T value, int rotate);
+
+template <typename T, typename = cpp::enable_if_t<cpp::is_unsigned_v<T>>>
+[[nodiscard]] LIBC_INLINE constexpr T rotl(T value, int rotate) {
+  constexpr unsigned N = cpp::numeric_limits<T>::digits;
+  rotate = rotate % N;
+  if (!rotate)
+    return value;
+  if (rotate < 0)
+    return cpp::rotr(value, -rotate);
+  return (value << rotate) | (value >> (N - rotate));
```

Comment:
@nickdesaulniers recommended a citation to "Safe, Efficient, and Portable Rotate in C/C++"
  // https://blog.regehr.org/archives/1063 in previous `__support/bit.h`. Maybe move it here?